### PR TITLE
Turbo version jni registry

### PR DESCRIPTION
**Pull request by karol-bisztyga**
This code reduces number of JNI searches to one per specific class and one per certain method

### CODE REVIEW

File: android/src/main/cpp/JNIRegistry.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    jclass classPtr = nullptr;
+    jmethodID methodPtr = nullptr;
+    // check if class is already stored, if not, try to obtain it
+    auto classIt = this->classes.find(className);
+    if (classIt != this->classes.end()) { // found
+        classPtr = classIt->second;
+    } else { // not found
+        classPtr = currentEnv->FindClass(className.c_str());
+        // register obtained class
+        classPtr = (jclass)currentEnv->NewGlobalRef(classPtr);
```
Comment (Szymon20000) [inline android/src/main/cpp/JNIRegistry.cpp]: I think we also have to delete this ref somewhere.

File: android/src/main/cpp/headers/JNIRegistry.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+enum class JNIMethodMode {
+    standard_method,
+    static_method,
+};
+
+class JNIRegistry {
+    JNIEnv* env;
+    // keys are class names
+    std::unordered_map<std::string, jclass> classes;
```
Comment (Szymon20000) [inline android/src/main/cpp/headers/JNIRegistry.h]: I think it's better to have all methods and classes that we use in arrays. It's a really important part of Reanimated.

Comment (karol-bisztyga) [reply android/src/main/cpp/headers/JNIRegistry.h]: fixed e0fd007

File: android/src/main/cpp/JNIRegistry.cpp (hunk 1 above)
Comment (karol-bisztyga) [reply android/src/main/cpp/JNIRegistry.cpp]: fixed ac5c74f

**Pull request merged**